### PR TITLE
feat(appset): Policies create-only, create-update, sync per ApplicationSet

### LINKED ISSUE #11073: Allow the Applicationset policy to be set on ApplicationSet object instead of global flag
# Summary

The [policy](https://argocd-applicationset.readthedocs.io/en/stable/Controlling-Resource-Modification/#policy-create-only-prevent-applicationset-controller-from-modifying-or-deleting-applications) is set per whole controller, while it should be set per ApplicationSet object. 

# Motivation

We use have staging and dev environments on the same k8s cluster separated by namespace. For staging it's not allowed to override Application parameters and applicationset `policy=create-update` should be applied. For dev namespaces it's totally fine to edit application parameters from cli or gui - for example to switch target branch to simplify testing of changes in helm charts. In that case we need `policy=create-only` to be set. 

# Proposal

default policy field can be overridden by spec of  ApplicationSet object.

### PR DESCRIPTION
**Pull request by speedfl**
Currently `create-only`, `create-update`, `sync` policies are only part of the controller launch parameters which does not allow to selectively disable deletion/updates.

## Motivations

- Provide a way to temporarily to deactivate the Applications update (ie: Disable AutoSync on one of the generated Applications) without impacting all projects/teams

- Provide a way to deactivate Applications deletion on a set of ApplicationSet and enable it on other without several instances of the ApplicationSet controller

Signed-off-by: gmuselli <geoffrey.muselli@gmail.com>

Closes #11073

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Note: I am not able to remove the kubebuilder annotations: https://github.com/go-swagger/go-swagger/issues/2687

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [X] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by speedfl** [comment]
@crenshaw-dev could you please help me on this ? I am not sure to understand what should I do. 
I guess there is something related to the [rbac configuration](https://argo-cd.readthedocs.io/en/stable/operator-manual/rbac/) but I don't know where and how should I perform my changes. If you maybe have some examples

**Comment by crenshaw-dev** [comment]
@speedfl you're asking for clarification in reference to my suggested change in policy override behavior? Or is your question regarding something else?

**Comment by speedfl** [comment]
@crenshaw-dev sorry it was not really clear.
I was asking for clarification in reference to your suggested change in policy override behavior.

**Comment by speedfl** [comment]
@crenshaw-dev I added the `--allow-policy-override` which is disabled by default if `--policy` is specified.

Added as well unit tests.
Integration tests are ok Locally

**Comment by crenshaw-dev** [comment]
@speedfl can you rebase?

File: docs/operator-manual/applicationset/Controlling-Resource-Modification.md
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Controlling-Resource-Modification.md:36]: I need to read the PR more, but why can't `syncPolicy.applicationsSync=create-only` prevent Application deletion?

Comment (speedfl) [reply docs/operator-manual/applicationset/Controlling-Resource-Modification.md:36]: It prevents the deletion in case the result of the generator removes an application. But from what I saw in the code the deletion of the applicationset deletes the application:

https://github.com/argoproj/argo-cd/blob/master/applicationset/controllers/applicationset_controller.go#L195

**Comment by speedfl** [comment]
@crenshaw-dev I see that `create-delete` sync policy has been added. I added one e2e and one unit test for this new use case

**Comment by speedfl** [comment]
Integration test are failing because of `TestSyncOptionValidateFalse` which also fails on master.

Pending to have it solved

image

**Comment by crenshaw-dev** [comment]
@speedfl fails on master locally or in CI? Could be an environment issue.

**Comment by speedfl** [comment]
@crenshaw-dev this morning everything worked. The CI was green.

**Comment by crenshaw-dev** [comment]
@speedfl I'm not going to have review time for this in the 2.7 cycle. I'm asking Ishita for help, maybe she can give it a first pass.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
@speedfl Added my comments.

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (speedfl) [reply pkg/apis/application/v1alpha1/applicationset_types.go:113]: I added it in the kubebuilder comment but not in the description. Fixing it

File: applicationset/utils/policy.go
Comment (speedfl) [reply applicationset/utils/policy.go:21]: The name of the method or parameter may be misleading I guess. It does not default on the SyncPolicy `sync` but on the controller policy.

Changing the parameter from `defaultPolicy` to `controllerPolicy`

File: applicationset/utils/policy.go
Comment (speedfl) [reply applicationset/utils/policy.go:14]: So you mean doing this in the applicationset_controller cmd

https://github.com/argoproj/argo-cd/pull/11462/files#diff-1ee9bd883626e217e79b2537de12484dd8f8e5a0d61cc792b56a86a9d9576b5b

```golang
var policyObj appv1alpha1.ApplicationsSyncPolicy
if policy == "" {
	policyObj = appv1alpha1.ApplicationsSyncPolicySync
} else {
	var exists bool
	policyObj, exists = utils.Policies[policy]
	if !exists {
		log.Info("Policy value can be: sync, create-only, create-update, create-delete. Default value: sync")
		os.Exit(1)
	}
}
```

I cannot force it in the cmd default value as the default value of `ARGOCD_APPLICATIONSET_CONTROLLER_ALLOW_POLICY_OVERRIDE` is based on the value `""` for the policy

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
I tested the PR and it works as expected. @speedfl could you rebase the PR and resolve conflicts?

**Comment by ishitasequeira** [comment]
@speedfl the lint error is fixed. Let's rebase this PR.

**Comment by speedfl** [comment]
Hey @ishitasequeira  I'll work on it tomorrow morning ! 🙂

**Comment by speedfl** [comment]
Hey @ishitasequeira do you know why I am geting error on `TestNamespacedGetLogsAllowSwitchOnNS` while I did not touched it ?

**Comment by ishitasequeira** [comment]
@speedfl AFAIK, It is a flaky test.. The solution for it is to re-trigger the test for now..

**Comment by speedfl** [comment]
Another test fail. Is there a way to relaunch the actions without doing a commit ?

**Comment by speedfl** [comment]
I run the test locally and it is working :/
[TestNamespacedGetLogsAllowSwitchOnNS.log](https://github.com/argoproj/argo-cd/files/11382044/TestNamespacedGetLogsAllowSwitchOnNS.log)
[TestNamespacedGetLogsAllowSwitchOnNS.run.log](https://github.com/argoproj/argo-cd/files/11382045/TestNamespacedGetLogsAllowSwitchOnNS.run.log)

**Comment by speedfl** [comment]
@ishitasequeira seems working now :)

**Comment by speedfl** [comment]
@ishitasequeira can we try to merge it now everything is green ?

File: applicationset/utils/policy.go
Comment (ishitasequeira) [reply applicationset/utils/policy.go:21]: Wondering if we want to change this to `controllerPolicy` as discussed in the comments.

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (ishitasequeira) [reply pkg/apis/application/v1alpha1/applicationset_types.go:102]: change the constants to `ApplicationsSyncPolicyCreateOnly`, `ApplicationsSyncPolicyCreateUpdate` and so on..

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
@speedfl , It seems the previously addressed comments have not been addressed anymore after the rebase. Might want to take a look again at the comments.

**Comment by speedfl** [comment]
@ishitasequeira should be good now. Sorry for that issue

**Comment by speedfl** [comment]
All good now!

**Comment by speedfl** [comment]
Hi @ishitasequeira any ETA for this one ?
We could set the milestones to 2.8 what do you think ?

File: cmd/argocd-applicationset-controller/commands/applicationset_controller.go
Comment (ishitasequeira) [inline cmd/argocd-applicationset-controller/commands/applicationset_controller.go:58]: nit: formatting

File: cmd/argocd-applicationset-controller/commands/applicationset_controller.go
Comment (ishitasequeira) [reply cmd/argocd-applicationset-controller/commands/applicationset_controller.go:96]: nit: ` '' (sync)"` -> `default value: sync`

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (ishitasequeira) [reply pkg/apis/application/v1alpha1/applicationset_types.go:118]: I think we missed adding `create-delete` here.

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (ishitasequeira) [inline pkg/apis/application/v1alpha1/applicationset_types.go:94]: Let's add a statement saying, "If no ApplicationsSyncPolicy is defined, it defaults it to `sync`". WDYT?

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
Thanks @speedfl, The overall PR looks good. Just minor nits.

File: pkg/apis/application/v1alpha1/applicationset_types.go
Comment (speedfl) [reply pkg/apis/application/v1alpha1/applicationset_types.go:94]: Better you are right

**Comment by ishitasequeira** [comment]
> Hi @ishitasequeira any ETA for this one ? We could set the milestones to 2.8 what do you think ?

@speedfl I just left last few comments (nits).. Once that is addressed, we would need an approver for second review to get it merged. We should be able to make it to 2.8 milestone though..

cc: @crenshaw-dev

**Comment by speedfl** [comment]
@ishitasequeira all done :)

File: docs/operator-manual/applicationset/Controlling-Resource-Modification.md
Comment (ishitasequeira) [inline docs/operator-manual/applicationset/Controlling-Resource-Modification.md:37]: I think I missed this, should we add `sync` here as an option as well?

**Review by ishitasequeira — APPROVED** [review]
LGTM!! Just added a quick documentation change that I missed out earlier.

@crenshaw-dev could you take another look at the PR?

File: docs/operator-manual/applicationset/Controlling-Resource-Modification.md
Comment (speedfl) [reply docs/operator-manual/applicationset/Controlling-Resource-Modification.md:37]: I added the policy sync.

**Review by crenshaw-dev — APPROVED** [review]
lfg, thanks as always @speedfl!

**Pull request merged**
